src/Microsoft.Azure.WebJobs.Host/Loggers/FunctionInstanceLogger.cs

```diff
@@ -24,54 +24,79 @@ public FunctionInstanceLogger(ILoggerFactory loggerFactory)
-        }
 
-        private void Log(LogLevel level, FunctionDescriptor descriptor, string message, params object[] args)
-        {
-            ILogger logger = _loggerFactory?.CreateLogger(LogCategories.CreateFunctionCategory(descriptor.LogName));
-            logger?.Log(level, 0, message, args);
+            logger.LogInformation(messageTemplate, templateValues);
         }
 
         public Task LogFunctionCompletedAsync(FunctionCompletedMessage message, CancellationToken cancellationToken)
         {
-            if (message.Succeeded)
-            {
-                string traceMessage = string.Format(CultureInfo.InvariantCulture, "Executed '{0}' (Succeeded, Id={1})", message.Function.ShortName, message.FunctionInstanceId);
-                Log(LogLevel.Information, message.Function, message.FunctionInstanceId, traceMessage);
-            }
-            else
-            {
-                string traceMessage = string.Format(CultureInfo.InvariantCulture, "Executed '{0}' (Failed, Id={1})", message.Function.ShortName, message.FunctionInstanceId);
-                Log(LogLevel.Error, message.Function, message.FunctionInstanceId, traceMessage, message.Failure.Exception);
-            }
+            var logger = GetOrCreateLogger(message.Function);
+            Log.FunctionCompleted(
+                logger, 
+                message.Function.ShortName, 
+                message.Succeeded ? "Succeeded" : "Failed", 
+                message.FunctionInstanceId, 
+                message.Failure?.Exception);
 
             return Task.CompletedTask;
         }
 
-        private void Log(LogLevel level, FunctionDescriptor descriptor, Guid functionId, string message, Exception exception = null)
+        public Task DeleteLogFunctionStartedAsync(string startedMessageId, CancellationToken cancellationToken)
         {
-            ILogger logger = _loggerFactory?.CreateLogger(LogCategories.CreateFunctionCategory(descriptor.LogName));
-            logger?.Log(level, 0, message, exception, (s, e) => s);
+            return Task.CompletedTask;
         }
 
-        public Task DeleteLogFunctionStartedAsync(string startedMessageId, CancellationToken cancellationToken)
+        private ILogger GetOrCreateLogger(FunctionDescriptor descriptor)
         {
-            return Task.CompletedTask;
+            return _cachedLoggers.GetOrAdd(descriptor.FullName, _ =>
+            {
+                return _loggerFactory?.CreateLogger(LogCategories.CreateFunctionCategory(descriptor.LogName));
+            });
+        }
+
+        private static class Log
+        {
+            private static readonly Action<ILogger, string, string, Guid, Exception> _functionStarted =
+                LoggerMessage.Define<string, string, Guid>(
+                    LogLevel.Information,
+                    new EventId(1, nameof(FunctionStarted)),
+                    "Executing '{name}' (Reason='{reason}', Id={id})");
```

Comment:
I use `functionName` elsewhere when doing structured logging -- let's make that consistent with these. And maybe make it `invocationId`? These keys are available to loggers now so making them descriptive can be helpful.

Reply:
Done